testng-core-api/src/main/java/org/testng/IParameterInfo.java

```diff
@@ -0,0 +1,36 @@
+package org.testng;
+
+/** Represents the ability to retrieve the parameters associated with a factory method. */
+public interface IParameterInfo {
+
+  /** @return - The actual instance associated with a factory method */
+  Object getInstance();
+
+  /**
+   * @return - The actual index of instance associated with a factory method. This index has a 1:1
+   *     correspondence with what were specified via the <code>indices</code> attribute of the
+   *     <code>@Factory</code> annotation. For e.g., lets say you specified the indices to the "1"
+   *     and your factory returned 4 instances, then the instance on which this method is invoked
+   *     would have the value as "1".
+   */
+  int getIndex();
+
+  /**
+   * @return - returns an index which indicates the running position in the array of test class
+   *     instances that were produced by a <code>@Factory</code> annotated constructor or static
+   *     method. For e.g., lets say your <code>@Factory</code> method returned 4 instances, then
+   *     each of the invocations to this method would return a value from <code>0</code> to <code>3
+   *     </code>
+   */
+  int currentIndex();
```

Comment:
The names between `getIndex` and `getCurrentIndex` are confusing.
What you mean here is that the index is a couple of 2 int (the running index related to the index from the dataset of the factory).

Reply:
I am also in confusion as to how to name it.

`getIndex()` has a 1:1 correlation between the indices specified. If not is specified then its value is equivalent to the value produced by `currentIndex()`
`currentIndex()` on the other has is a simple running counter, that tells you the exact `ith` iteration when this object was created. 

Please let me know if you have any better way of expressing this.